# HW 1: Minimal Pair with Implicit Causality
### COSC 426: Fall 2025, Colgate University

Use this notebook to load in the results from your experiment, display them, and answer the question(s). Feel free to add as many code and markdown chunks as you would like in each of the sub-sections. 

**If you use any external resources (e.g., code snippets, reference articles), please cite them in comments or text!**

## 1. How does the model's NP1 bias change with verb frequency? 

### My Plan
I will use minimum pair with comparison by frequency column.

For frequency column, I'll get the median frequency of the verbs, then will label 
pair as `frequent` if the verb's frequency is greater than the median and `not frequent` if it is less than the median.

The sentence structures I will use for this task are

1. "Marry `verb` John becuase she ..." // `NP1` biased
2. "Marry `verb` John becuase he ..."  // `NP2` biased

For comparison("un/expected") column, 

- If the verb's NP1 > NP2, then the sentence with "she" is expected  
- If the verb's NP2 > NP1, then the sentence with "he" is expected

### Setup and data loading
In this section, load the data, import any packages or files, and define any functions you need.

In [1]:
import pandas as pd
import utils

In [ ]:
original_df = utils.read_data(fpath = "data/data.xlsx", sheet_name = "data96subjects", skiprows = 1)

original_df.to_csv("data/original_data.tsv", sep="\t", index=False) # Save as TSV. Not necessary

df = utils.generate_dataset(original_df) # Generate the dataset
df.to_csv("data/data.tsv", sep="\t", index=False) # Save as TSV.

### Results from the experiment
In this section, display the results required to answer your question. This can be in the format of a dataframe with just the required rows, or as a plot. 

In [7]:
result = pd.read_csv("results/results.tsv", sep="\t", index_col=0)
result.head()

model bias   frequency       acc      diff  expected  \
0  distilbert/distilgpt2  NP1    frequent  0.000000 -0.108063  0.020622   
1  distilbert/distilgpt2  NP1  infrequent  0.000000 -0.123471  0.012061   
2  distilbert/distilgpt2  NP2    frequent  0.953488  0.095766  0.111352   
3  distilbert/distilgpt2  NP2  infrequent  0.912281  0.112739  0.139033   

   unexpected  macrodiff  
0    0.128685  -0.108063  
1    0.135532  -0.123471  
2    0.015586   0.095766  
3    0.026294   0.112739

### Discussion
In this section, include a discussion of your results and how it pertains to the question(s) you picked. This section will likely only have markdown chunks

The model's NP1 bias is slightly stronger when the verb is infrequent compared to when it's frequent. 

This is clear when looking at the diff values for the rows with an NP1 bias, where infrequent verbs show a score of -0.123, a more negative and thus stronger indicator of bias than the -0.108 score for frequent verbs. 

The same pattern holds true for the model's NP2 bias. The data shows a diff of 0.112 for infrequent NP2-biased verbs compared to 0.095 for frequent ones. Since a larger positive value signifies a stronger NP2 bias, this affirms the trend. So both biases are more pronounced when the model encounters less common verbs.

---

## 2. Do activity verbs have a weaker NP1 bias than psychological verbs?

I would have used analogous logic to build the dataset, creating minimal pairs that are compared by verb category.

For the 'category' column, I would have labeled verbs with the semantic codes 'AgPat' or 'AgEvo' as 'activity', and those with 'ExpStim' or 'StimExp' as 'psychological'.

The sentence structures I would have used for this task are:
1.  "Mary `verb` John because she..."
2.  "Mary `verb` John because he..."

For the 'comparison' column, a sentence would be labeled 'expected' or 'unexpected' based on the verb's bias:
* If a verb's NP1 score is greater than its NP2 score, the sentence with "she" is considered expected, and the one with "he" is unexpected.
* If a verb's NP2 score is greater than its NP1 score, the sentence with "he" is considered expected, and the one with "she" is unexpected.